make private fields final where possible

oreissig opened the pull request:
As discussed in PR #282 I repropose the changes from there in individual pull requests.

jwagenleitner commented on src/main/org/codehaus/groovy/control/MultipleCompilationErrorsException.java:
```diff
@@ -27,7 +27,7 @@
         CompilationFailedException {
     
-    protected ErrorCollector collector;
+    protected final ErrorCollector collector;
```
There could be subclasses that set this field.

jwagenleitner commented on subprojects/groovy-sql/src/main/java/groovy/sql/BatchingStatementWrapper.java:
```diff
@@ -34,10 +34,10 @@
     protected int batchSize;
     protected int batchCount;
-    protected Logger log;
+    protected final Logger log;
```
While it may not be likely, there could be some subclasses that set this so I think it might be good to avoid changing `protected` fields.

jwagenleitner commented:
With the exception of a few changes to protected fields, it looks good to me.  With such a large change I would be curious what others think.  Hopefully get some to comment here.

shils commented on src/main/groovy/grape/GrapeIvy.groovy:
```diff
@@ -76,11 +76,11 @@ class GrapeIvy implements GrapeEngine {
     Set<String> downloadedArtifacts
     // weak hash map so we don't leak loaders directly
-    Map<ClassLoader, Set<IvyGrabRecord>> loadedDeps = new WeakHashMap<ClassLoader, Set<IvyGrabRecord>>()
+    final Map<ClassLoader, Set<IvyGrabRecord>> loadedDeps = new WeakHashMap<ClassLoader, Set<IvyGrabRecord>>()
```
making these fields final will break binary compatibility since their synthetic public setters won't be generated

blackdrag commented:
I agree with @jwagenleitner. We cannot merge this pull request as it is right now.  Frankly I prefer doing a change to final only if I mean to have it final. If for example all but one field is mutated from different places, then making that one field private does not really serve any purpose. Anyway... this should be done only for private fields. Fields of private inner classes are probably save, at least if not written in Groovy.

graemerocher commented:
Yeah this has huge implications to Groovy libraries and could break existing code

oreissig commented:
I'll revert the changes to any non-private fields. I agree that we should not change any signiture that's available to others, even if they are unlikely to rely on that.

oreissig commented:
Done reverting.

I'm not quite sure about the argument with Groovy ignoring the private flag, making this change potentially breaking. By this logic, every change that does not merely add new classes is a breaking change. Maybe we could treak the groovy.* package different than org.codehaus.groovy.*, with the former employing an even stronger compatibility contract?

blackdrag commented:
Groovy does not ignore private, but private has slightly different semantics in Groovy if compared to Java. I mentioned inner classes only because of some quirks there that may have a different effect from what you expect.

blackdrag commented:
If you want to change something that has been available in a non-private context, then we have to discuss each and every case.

oreissig commented:
So you are basically referring to [this issue](https://issues.apache.org/jira/browse/GROOVY-1591)? From what I am reading the "slightly different semantics" means "private is respected with `@CompileStatic` and accessible as if it were public for dynamic Groovy".

Do you mean this change is only safe for private inner classes?

jwagenleitner commented on subprojects/groovy-macro/src/main/groovy/org/codehaus/groovy/macro/matcher/internal/MatchingConstraintsBuilder.groovy:
```diff
@@ -24,7 +24,7 @@ import org.codehaus.groovy.macro.matcher.TreeContext
 
 class MatchingConstraintsBuilder {
-    Set<String> placeholders = new LinkedHashSet<>()
+    final Set<String> placeholders = new LinkedHashSet<>()
```
Since this is a Groovy object this is changing a public property to final and I don't think should be done.

jwagenleitner commented on src/main/org/codehaus/groovy/ast/expr/GStringExpression.java:
```diff
@@ -32,13 +32,15 @@
         this.verbatimText = verbatimText;
         super.setType(ClassHelper.GSTRING_TYPE);
+        this.strings = new ArrayList<ConstantExpression>();
+        this.values = new ArrayList<Expression>();
```
This would cause a breaking change if any code relied on `#getStrings` or `#getValues` returning `null`.  I would avoid this change.

jwagenleitner commented on src/main/org/codehaus/groovy/runtime/ProcessGroovyMethods.java:
```diff
@@ -433,7 +433,7 @@ public static Process or(final Process left, final Process right) throws IOExcep
      */
     protected static class ProcessRunner implements Runnable {
-        Process process;
+        final Process process;
```
This could be a breaking change.

oreissig commented:
Thanks for the feedback. I reverted the changes I accidentally made to the Groovy class and changed the expression's behaviour to initialize its fields with null instead of empty list.

oreissig commented:
Good thing there are test :) The `getString()` or `getValues()` methods of `GStringExpression` are actually assumed to not return null.

jwagenleitner commented:
I missed that in the original code those were assigned where they were declared and that you had moved it down to the constructor.

oreissig commented:
I rebased the changes from master and squashed the corrections.

oreissig commented:
What's the future of this PR?
Maintaining a feature branch with such widespread changes to be conflict free is not rewarding if the outcome is unclear.
Please decide if it should be abandoned at all, if I should just fix the merge conflicts, or if you see the possibility to merge a subset of changes.

blackdrag commented:
I think this pull request is too big. Try splitting it in bunches of <20 files, for example on a per subproject base would already help. And then maybe package path based for src/main. This will allow more people to do the review, and avoids maintenance if more merge conflicts appear.

oreissig commented:
The discussion raised concernes that we should avoid this kind of change even for private fields. If so, then I don't need to continue. Otherwise I'm fine with rebasing and splitting the PR.

blackdrag commented:
I think it is ok to do for private fields

paulk-asert commented:
I merged this PR but tweaked to exclude changes to these files:
```
src/main/groovy/util/ClosureComparator.java
src/main/org/codehaus/groovy/antlr/LexerFrame.java
src/main/org/codehaus/groovy/control/CompilePhase.java
src/main/org/codehaus/groovy/runtime/ProcessGroovyMethods.java
src/main/org/codehaus/groovy/runtime/metaclass/MetaMethodIndex.java
subprojects/groovy-groovydoc/src/main/java/org/codehaus/groovy/tools/groovydoc/ClasspathResourceManager.java
subprojects/groovy-groovydoc/src/main/java/org/codehaus/groovy/tools/groovydoc/MockOutputTool.java
subprojects/groovy-groovydoc/src/main/java/org/codehaus/groovy/tools/groovydoc/SimpleGroovyExecutableMemberDoc.java
subprojects/groovy-json/src/main/java/groovy/json/internal/ValueList.java
subprojects/groovy-sql/src/main/java/groovy/sql/CallResultSet.java
```
For those files a package-private field was made final in a non-private class. These all break binary compatibility. For some of the internal `org.codehaus.groovy` classes we _might_ decide to do that but it should be on a case-by-case basis, not in the same commit as the non-controversial changes.
